### ISSUE TITLE
[Tracker] ImagePullBack Error is occuring in all clusters

### ISSUE THREAD

**Issue by harshad16** [thoth-station/thoth-application#1996]
**Describe the bug**
We have our services deployed using different openshift objects like deployment, deployment config, cronjob, Argo workflows, etc.
recently, the pods corresponding to these objects are throwing an **imagePullBack** error. 
The imagestream is available and has the **lookupPolicy** local **true**. 
It seems the pods are not able to lookup the imagestream and are looking for the image in docker.io.

Related Issues:
https://github.com/thoth-station/thoth-application/pull/1995

Alternates:
we can shift away from imagestream and start using images directly.

**Comment by harshad16** [comment]
This was discussed in the tech talk.
the reference of imagestream https://docs.openshift.com/container-platform/4.8/openshift_images/using-imagestreams-with-kube-resources.html, has a solution for the deployment, which seems to be functioning.

However the issue persists in the argo workflows as well.

**Comment by harshad16** [comment]
![backend-stage](https://user-images.githubusercontent.com/14028058/135632719-bab02421-3fec-42a1-ab58-efbc7afd88d0.png)
![integration-test](https://user-images.githubusercontent.com/14028058/135632720-62d82b70-10c8-435a-b322-61073327a0d9.png)
![cve-image](https://user-images.githubusercontent.com/14028058/135632722-e153d270-eb41-456b-b16a-d51f08fd9cd1.png)

**Comment by goern** [comment]
This feels like an upstream OpenShift bug to me, shall we have a look how to handle image versioning from outside openshift (aka ArgoCD) ?

/triage accepted
/assign @harshad16

**Comment by harshad16** [comment]
https://chat.google.com/room/AAAAVjnVXFk/p-OrI5ywZBo

**Comment by goern** [comment]
Same for https://console-openshift-console.apps.smaug.na.operate-first.cloud/k8s/ns/thoth-bots-prod/deploymentconfigs/sefkhet-abwy-webhook-receiver/pods

**Comment by harshad16** [comment]
The workaround seems to be adding an absolute path.
using the kustomize feature, we would apply the absolute path of the image. 
in this way, we would be able to keep on using the imagestream.


The snippet of the solution in Kustomization.yaml:

For workflows:
```
  - patch: |-
      - op: replace
        path: /spec/templates/0/container/image
        value: image-registry.openshift-image-registry.svc:5000/thoth-backend-stage/user-api
    target:
      group: argoproj.io
      version: v1alpha1
      kind: WorkflowTemplate
      name: advise
```
For others like deployment, etc:
```
 images:
  - name: user-api
    newName: image-registry.openshift-image-registry.svc:5000/thoth-frontend-stage/user-api
    newTag: latest
```

**Comment by goern** [comment]
any update in this?

**Comment by codificat** [comment]
The fix for [BZ 2000216](https://bugzilla.redhat.com/show_bug.cgi?id=2000216) reached OCP 4.9 in 4.9.7 via [BZ 2018442](https://bugzilla.redhat.com/show_bug.cgi?id=2018442): [RHBA-2021:4579](https://access.redhat.com/errata/RHBA-2021:4579).

While the workaround is working, we should probably be good to consider reverting it moving forward.

**Comment by goern** [comment]
do we have a status which op1st clusters run which version of OCP?

is it safe to close this issue?

**Comment by codificat** [comment]
> do we have a status which op1st clusters run which version of OCP?

AFAIK smaug is at 4.8.23. Recently (23/Feb) a planned upgrade to 4.9.21 was postponed:

> Hello folks, it has come to our attention there yet still remain pesky deprecated apis on the Smaug cluster that need to be cleaned up before we can proceed with the upgrade. As such we are postponing the upgrade to 4.9.21 for now, we will provide an update to a new date at a later time.

However, 4.8.23 already contains a fix for this, which was included in 4.8.22 via [BZ 2020644](https://bugzilla.redhat.com/show_bug.cgi?id=2020644): [RHBA-2021:4830](https://access.redhat.com/errata/RHBA-2021:4830)

So, it should be ok to revert the workaround in smaug-based resources.

Balrog reports Kubernetes Version: v1.21.1+d8043e1 which I believe corresponds to OCP 4.8.12, but op1st admins can confirm better.

> is it safe to close this issue?

We could open another one to revert the workaround, but it's probably better to just keep using this one to track it.

Should we test reverting it on some smaug resource to verify it works?

**Comment by harshad16** [comment]
> Should we test reverting it on some smaug resource to verify it works?

Tried it with other components, but some how it doesn't work in the upgraded version.

**Comment by codificat** [comment]
It was mentioned in today's meeting that 4.10 is really supposed to have fixed this, maybe we will get to undo the workaround later

**Comment by harshad16** [comment]
Closing the issue 
as it was fixed  and now the cluster that had these issues are not used anymore.
